processor/src/main/java/org/mapstruct/ap/internal/model/beanmapping/TargetReference.java

```diff
@@ -56,55 +58,70 @@
-        private Mapping mapping;
         private Method method;
         private FormattingMessager messager;
         private TypeFactory typeFactory;
 
+        // mapping parameters
+        private boolean isReversed = false;
+        private boolean isIgnored = false;
+        private String targetName = null;
+        private AnnotationMirror annotationMirror = null;
+        private AnnotationValue targetAnnotationValue = null;
+        private AnnotationValue sourceAnnotationValue = null;
+        private Method templateMethod = null;
         /**
          * During {@link #getTargetEntries(Type, String[])} an error can occur. However, we are invoking
          * that multiple times because the first entry can also be the name of the parameter. Therefore we keep
          * the error message until the end when we can report it.
          */
         private MappingErrorMessage errorMessage;
 
-        public BuilderFromTargetMapping messager(FormattingMessager messager) {
+        public Builder messager(FormattingMessager messager) {
             this.messager = messager;
             return this;
         }
 
-        public BuilderFromTargetMapping mapping(Mapping mapping) {
-            this.mapping = mapping;
+        public Builder mapping(Mapping mapping) {
+            if ( mapping.getInheritContext() != null ) {
+                this.isReversed = mapping.getInheritContext().isReversed();
+                this.templateMethod = mapping.getInheritContext().getInheritedFromMethod();
+            }
+            this.isIgnored = mapping.isIgnored();
+            this.targetName = mapping.getTargetName();
+            this.annotationMirror = mapping.getMirror();
+            this.targetAnnotationValue = mapping.getTargetAnnotationValue();
+            this.sourceAnnotationValue = mapping.getSourceAnnotationValue();
             return this;
         }
 
-        public BuilderFromTargetMapping method(Method method) {
-            this.method = method;
+        public Builder typeFactory(TypeFactory typeFactory) {
+            this.typeFactory = typeFactory;
             return this;
         }
 
-        public BuilderFromTargetMapping typeFactory(TypeFactory typeFactory) {
-            this.typeFactory = typeFactory;
+        public Builder method(Method method) {
+            this.method = method;
             return this;
         }
 
         public TargetReference build() {
 
-            boolean isInverse = mapping.getInheritContext() != null && mapping.getInheritContext().isReversed();
+            Objects.requireNonNull( method );
+            Objects.requireNonNull( typeFactory );
 
-            String targetName = mapping.getTargetName();
             if ( targetName == null ) {
                 return null;
             }
 
             String targetNameTrimmed = targetName.trim();
-            if ( !targetName.equals( targetNameTrimmed ) ) {
+            if ( !targetName.equals( targetNameTrimmed ) && messager != null ) {
```

Comment:
Same here `messager` should never be `null`. Otherwise we can't report.

Reply:
intentional.. See above

Reply:
resolved